New concept: Optional

josealonso opened the pull request:
Write explanation for the Optional concept, as well as the test file, the stub implementation, the reference solution and other required files. Some files must be modified. This is a draft.

This solves #2555 


<!-- DO NOT EDIT BELOW THIS LINE! -->
---

Reviewer Resources:

[Track Policies](https://github.com/exercism/java/blob/main/POLICIES.md#event-checklist)

josealonso commented:
I just committed a first draft. Don't worry about the linting errors, I will sort them out during this week.
I'd like to receive feedback about the reference solution and the test file, the bulk of the work is done.
- Is there an exercise for the `generics` concept ? That is a prerequisite for this concept. 
- If that concept is not explained on the Java track, I can adapt it from other track.

@kahgoh @

kahgoh commented:
I hope to have a look at the draft soon.

In regards to the `generics` concept, it is covered as [generics type](https://exercism.org/tracks/java/concepts/generic-types).

josealonso commented:
> Thanks @josealonso! I plan to have a look at the draft soon.
> 
> In regards to the `generics` concept, it is covered as [generics type](https://exercism.org/tracks/java/concepts/generic-types).

Thanks for the link to the generic-types exercise!! Shame on me

josealonso commented:
Which icon should I choose?

kahgoh commented:
Did you mean the icon for the exercise? They get added to [exercism/icons](https://github.com/exercism/website-icons/) repository, though that can be done after the exercise is added. We'll have to make one, but I think it will be the same icon as the original Tim From Marketing, but with a "2" in the top right corner. Compare [Wizards and Warriors](https://exercism.org/tracks/java/exercises/wizards-and-warriors) and [Wizards and Warriors 2](https://exercism.org/tracks/java/exercises/wizards-and-warriors-2) for an example.

kahgoh commented on exercises/concept/tim-from-marketing-2/.meta/src/reference/java/EmployeeService.java:
```diff
@@ -0,0 +1,66 @@
+
+
+class EmployeeService {
+  
```
To be honest, I didn't quite get the purpose of these two methods (`getAllTheEmployeesById` and `getEmployeeById`) especially as we're not expecting the students to implement these and I don't think its clear when we expect them to call them. 

Since the `Employee` have methods that return `Optional` objects, I'd suggest having the test pass `Employee` objects or  a `List` of them instead. 

Is it also worth having a task for where the student has to return an `Optional` as well?

josealonso replied:
I think this has been resolved.

kahgoh replied:
I had another thought - the `getEmployeeById` could also be left up to the student to implement. We already have the [`foreach` concept](https://exercism.org/tracks/java/concepts/foreach-loops), so they could use that to find the employee and that'll give them the chance to practice making an `Optional` object and there would be no need for streams.

josealonso replied:
@kahgoh, despite I spent many hours creating this exercise, I have to admit it's probably a bit complex for a learning exercise. This is what I propose:

- Leaving this one for a future practice exercise.
- Create a concept exercise for **Streams** first. And maybe another one for Suppliers or functional interfaces, based on [my article](https://blog.programwithjr.com/lambda-expressions-and-function-types) or a similar one.

kahgoh replied:
Hey @josealonso, it sounds like you've struggled with coming up with an exercise. Before switching, I was wondering ... what is making it complex or where do you see the complexity lie in making a learning exercise? I did think the exercise still needed work, but it was still something that could be built upon.

> Leaving this one for a future practice exercise.

Since the first Tim for Marketing is a learning exercise, I think Tim for Marketing 2 should be kept as a learning exercise (as opposed to a practice exercise). Happy to use it for a different concept (if we find an appropriate one later) or develop a similar practice exercise later (let's just give it a different story and name 😉).

> Create a concept exercise for Streams first.

Streams is definitely something we should cover! Although, I disagree that it has to come first. My concern with doing Streams first is that there are `Stream` operations that return an `Optional` (e.g. `findAny` and `findFirst`). Would we end up with a circular dependency? Or is there a way to we can do it in a way without circular dependency? I was thinking we could cover `Optional` on its own, outside  of `Stream`.

> And maybe another one for Suppliers or functional interfaces, based on [my article](https://blog.programwithjr.com/lambda-expressions-and-function-types) or a similar one.

Perhaps a bit related, I had also thought "Functional Interfaces, Lambdas and Method References" would make a good concept (though, not specifically `Suppliers`). This probably would be a pre-requisite to streams, but someone else is already working on "Functional Interfaces, Lambdas and Method References" though. I don't think this should stop you from switching to specifically Streams if you want to though.

josealonso replied:
I'll leave a longer answer tomorrow, but I just wanted to thank you for your support and clarify that my struggle is due to being my first exercise designed from scratch and also because I wasn't not very familiar with the Optional API.
I would cover more simple streams operations in a first Streams learning exercise, like map, filter and other simple operations.
@kahgoh

josealonso replied:
> I had another thought - the `getEmployeeById` could also be left up to the student to implement. We already have the [`foreach` concept](https://exercism.org/tracks/java/concepts/foreach-loops), so they could use that to find the employee and that'll give them the chance to practice making an `Optional` object (assuming they are looking through a `List<Employee>`) and there would be no need for streams in the class.

Are you saying I can implement the same methods without using streams ? I'm not sure that's doable. :thinking:

kahgoh replied:
Sorry I think I missed the way the exercise or test is set up at the moment. I had imagined the `EmployService` take in a `List<Employee>` in the constructor, so then the students could be tasked with finding the Employee by id. I imagine the stub would like this:

```java
class EmployeeService {
  public EmployeeService(List<Employee> employees) {
    throw new UnsupportedOperationException("Please implement the EmployeeService constructor");
  }

  public Optional<Employee> getEmployeeById(int id) {
     throw new UnsupportedOperationException("Please implement the EmployeeService.getEmployeeById() method");
  }
}
```

Students can still make a solution for this using `for` loop. 

Alternatively, I think passing the method a `List<Employee>` would also work. If keeping `getAllTheEmployeesById`, I think it might help to change it to `getAllEmployees` return a `List<Employee>` instead (I notice the `id` is also in `Employee` class, so it seems strange to need to "map" them to the list index).

josealonso replied:
> I was thinking the `EmployService` could take in a `List<Employee>` in the constructor, so then the students could be tasked with finding the Employee by id. I imagine the stub would like this:
> 
> ```java
> class EmployeeService {
>   public EmployeeService(List<Employee> employees) {
>     throw new UnsupportedOperationException("Please implement the EmployeeService constructor");
>   }
> 
>   public Optional<Employee> getEmployeeById(int id) {
>      throw new UnsupportedOperationException("Please implement the EmployeeService.getEmployeeById() method");
>   }
> }
> ```
> 
> Students can still make a solution for this using a `for` / `while` loop that goes through the list until it finds the Employee. Here's one rough way:
> 
> ```java
> for (Employee candidate: employeesList) {
>   if (Objects.equals(candidate.getId(), Optional.of(id)) {
>     return Optional.of(candidate);
>   }
> }
> 
> return Optional.empty();
> ```
> 
> Alternatively, I think passing the method a `List<Employee>` to the method would also work.

Thanks @kahgoh. I'll try to change it this weekend. I was stuck, because I designed this exercise to try to mimic the access to a database. But you're right, I made it too complex.

kahgoh commented on exercises/concept/tim-from-marketing-2/src/test/java/EmployeeServiceTest.java:
```diff
@@ -0,0 +1,101 @@
+                .isEqualTo(expectedListOfPrintedEmployeesNames);
+    }
+
+    @ParameterizedTest
```
There's a preference to avoid `Parameterized` tests as it makes it harder to do TDD (see #1293 for previous discussions around this).

josealonso commented:
> Did you mean the icon for the exercise? They get added to [exercism/icons](https://github.com/exercism/website-icons/) repository, though that can be done after the exercise is added. We'll have to make one, but I think it will be the same icon as the original Tim From Marketing, but with a "2" in the top right corner. Compare [Wizards and Warriors](https://exercism.org/tracks/java/exercises/wizards-and-warriors) and [Wizards and Warriors 2](https://exercism.org/tracks/java/exercises/wizards-and-warriors-2) for an example.

Who can make that icon? I'm sorry I don't know how to edit a simple icon :(

josealonso commented:
I don't understand why the java-test-runner failed, @kahgoh. It probably has to do with some missing configuration files,

kahgoh commented:
> > Did you mean the icon for the exercise? They get added to [exercism/icons](https://github.com/exercism/website-icons/) repository, though that can be done after the exercise is added. We'll have to make one, but I think it will be the same icon as the original Tim From Marketing, but with a "2" in the top right corner. Compare [Wizards and Warriors](https://exercism.org/tracks/java/exercises/wizards-and-warriors) and [Wizards and Warriors 2](https://exercism.org/tracks/java/exercises/wizards-and-warriors-2) for an example.
> 
> Who can make that icon? I'm sorry I don't know how to edit a simple icon :(

That's fine. We can do that later. For now, let's just work on the concept content.

kahgoh commented:
> I don't understand why the java-test-runner failed, @kahgoh. It probably has to do with some missing configuration files,

Its a bit hard to find, but this appears in the [log](https://github.com/exercism/java/actions/runs/13340322597/job/37263741182?pr=2913#step:3:69):

> tim-from-marketing-2: exemplar solution did not pass the tests

josealonso commented:
> 
> Are the tests passing locally for you?

Yes.

manumafe98 commented on exercises/concept/tim-from-marketing-2/.docs/hints.md:
```diff
@@ -0,0 +1,11 @@
+
+WIP
+
+
```
@josealonso I suggest you to use this [vscode extension](https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint) for markdown linting, you will be able to solve them easily

josealonso replied:
Thanks for the suggestion!! I use Intellij, there might be a similar plugin for this IDE, I guess.

josealonso replied:
I just installed it. Thanks @manumafe98!

manumafe98 commented on exercises/concept/tim-from-marketing-2/src/test/java/EmployeeServiceTest.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.ArrayList;
+import java.util.Optional;
+
+// Annotation in order to use @BeforeAll in a non-static method
```
Do you think the comments on this test add value? I think the code is pretty self explanatory (to see if you are going to delete them in the final version)

manumafe98 commented on exercises/concept/tim-from-marketing-2/.meta/design.md:
```diff
@@ -0,0 +1,47 @@
+
+This exercise could benefit from the following rules in the [analyzer]:
+
+- `actionable`: If the solution uses `null` in any method, encourage the student to use `Optional<T>` instead.
```
I would change this one to `essential` as I think is literally the core of Optionables

josealonso replied:
I agree. I did not know about `essential` actions, thanks.

manumafe98 commented:
Maybe this doc on [configlet](https://exercism.org/docs/building/configlet) could be useful for you as well to check

josealonso commented:
> Maybe this doc on [configlet](https://exercism.org/docs/building/configlet) could be useful for you as well to check

Thanks, I went through it already. I'll have to read it again.

kahgoh commented on exercises/concept/tim-from-marketing-2/src/test/java/EmployeeServiceTest.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.Optional;
+
+// Annotation in order to use @BeforeAll in a non-static method
+@TestInstance(Lifecycle.PER_CLASS)
```
I think this is causing the failure in the CI. It should work if you remove this and change the `@BeforeAll` on line 36 to `@BeforeEach`.

```suggestion
```

kahgoh commented on exercises/concept/tim-from-marketing-2/src/test/java/EmployeeServiceTest.java:
```diff
@@ -0,0 +1,75 @@
+        listOfEmployees.add(Optional.of(new Employee(2, "John", "Engineering")));
+    }
+
+    @BeforeAll
```
As mentioned in another comment, changing this should get the tests passing in the CI.

```suggestion
    @BeforeEach
```

kahgoh commented on exercises/concept/tim-from-marketing-2/src/test/java/EmployeeServiceTest.java:
```diff
@@ -0,0 +1,75 @@
+        listOfEmployees.add(Optional.of(new Employee(2, "John", "Engineering")));
+    }
+
+    @BeforeAll
```
As mentioned in another comment, changing this should get the tests passing in the CI.

```suggestion
    @BeforeEach
```

josealonso replied:
Many thanks, @kahgoh !! That makes sense, because it's working locally.

josealonso commented:
@kahgoh, can you please go through these files and tell me if that's the right approach ? Thanks :)

kahgoh commented on exercises/concept/tim-from-marketing-2/src/test/java/EmployeeServiceTest.java:
```diff
@@ -0,0 +1,84 @@
+        listOfEmployees.add(new Employee(1, "Tim", "Marketing"));
+        listOfEmployees.add(new Employee(2, "Joe", "Sales"));
+        listOfEmployees.add(new Employee(3, "Jane", "IT"));
+        listOfEmployees.add(new Employee(4, null, null));
```
I think there are a number of cases missing,  such as a `null` name and non-null department and non-null name and null department.

josealonso replied:
WIP

kahgoh commented on exercises/concept/tim-from-marketing-2/.meta/src/reference/java/EmployeeService.java:
I suspect not everyone knows what a "service" class is. I'd suggest using a term that is easier to understand by more people (e.g. EmployeeDatabase?).

josealonso replied:
Hey @kahgoh, I'm sorry I didn't get back to you earlier. I will try to apply all your suggestions during this week. Thanks for such a detailed feedback!

josealonso replied:
Renamed to **EmployeeDatabase**

kahgoh reviewed (commented):
Hey @josealonso, I have had a look through and I think the exercise is going in the right direction. 

In a couple of the comments I mention an [about.md](https://exercism.org/docs/building/tracks/concepts#h-file-about-md). The about.md is shown after they complete the exercise and would be something that will need to eventually be added for the concept.

josealonso commented on exercises/concept/tim-from-marketing-2/build.gradle:
```diff
@@ -0,0 +1,23 @@
+dependencies {
+    testImplementation platform("org.junit:junit-bom:5.10.0")
+    testImplementation "org.junit.jupiter:junit-jupiter"
+    testImplementation "org.assertj:assertj-core:3.25.1"
```
Resolved locally.

josealonso commented on config.json:
```diff
@@ -309,6 +309,20 @@
+        "name": "tim-from-marketing-2",
+        "uuid": "a6cfc286-8c62-4f5b-8e59-f6bfc4374092",
+        "concepts": [
+          "optional type"
```
Resolved locally.

josealonso commented on exercises/concept/tim-from-marketing-2/.meta/design.md:
```diff
@@ -0,0 +1,45 @@
+The goal of this exercise is to teach the student how to use the Optional API.
+We will use the most common methods: `ifPresent`, `orElse`, `ifPresentOrElse`, `orElseThrow`. 
+The `isPresent` and `get` methods are not presented, since they do not provide any value over an ordinary null check.
+Some methods of the Stream API are needed. This is a bit problematic, since they have not been explained in the current Java track.
```
Done.

josealonso commented on exercises/concept/tim-from-marketing-2/src/test/java/EmployeeServiceTest.java:
```diff
@@ -0,0 +1,84 @@
+
+    @BeforeEach
+    void setup() {
+        initList();
```
I couldn't do it like that, because I was mutating an element list in the null case. I will revisit it.

kahgoh replied:
I'm not quite sure where you are mutating the list element, but if a test needs different data, I'd suggest setting up the test data in the test methods instead.

josealonso commented on exercises/concept/tim-from-marketing-2/.docs/introduction.md:
```diff
@@ -0,0 +1,69 @@
+
+## Introduction
+
+The **Optional<T>** type was introduced in Java 8 as a way to indicate that a method will return an object of type T or an empty value. It is present in type signatures of many core Java methods. 
```
Done.

josealonso commented on exercises/concept/tim-from-marketing-2/.meta/src/reference/java/Employee.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.Objects;
+import java.util.Optional;
+
```
Done.

josealonso commented on exercises/concept/tim-from-marketing-2/.docs/introduction.md:
```diff
@@ -0,0 +1,69 @@
+## Introduction
+
+The **Optional<T>** type was introduced in Java 8 as a way to indicate that a method will return an object of type T or an empty value. It is present in type signatures of many core Java methods. 
+Before Java 8, developers had to implement null checks:
```
Done.

josealonso commented on exercises/concept/tim-from-marketing-2/.docs/introduction.md:
````diff
@@ -0,0 +1,69 @@
+    }
+}
+```
+
````
WIP

josealonso commented:
I will continue tomorrow.

kahgoh commented:
> I will continue tomorrow.

No worries @josealonso . There's no hurry for this one, so feel free to take longer if you need to

josealonso commented:
> > I will continue tomorrow.
> 
> No worries @josealonso . There's no hurry for this one, so feel free to take longer if you need to

- The value of `concepts.uuid` is a zero-length string. 
I thought it had to match the `exercises.uuid`, but it gave an error, so I left it empty. 
It's the only error in the configlet phase.

- The tests are failing to build even locally, it might be a configuration issue.

I addressed all the suggestions you made, except using List.of() instead of a function in the tests.

Please tell me how to fix the two errors mentioned above and fell free to suggest any improvements.

Thanks!

kahgoh commented:
I've just tried running the tests locally. The failing tests have the following messages:

```
EmployeeDatabaseTest > Retrieve employee details by id when department is null and name is not null FAILED
    org.opentest4j.AssertionFailedError:
    expected: "No employee found for id: 6"
     but was: "6 - Alice - null"
        at java.base@21.0.5/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
        at java.base@21.0.5/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
        at app//EmployeeDatabaseTest.retrieveEmployeeDetailsById_whenDepartmentIsNull(EmployeeDatabaseTest.java:92)
```

and 
```
EmployeeDatabaseTest > Retrieve employee details by id when the id does not exist FAILED
    org.opentest4j.AssertionFailedError:
    expected: "No employee found for id: 10"
     but was: "That id does not exist: 10"
        at java.base@21.0.5/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
        at java.base@21.0.5/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
        at app//EmployeeDatabaseTest.retrieveEmployeeDetailsById_forANonExistingId(EmployeeDatabaseTest.java:100)
```

I ran them against the `.meta/src/reference` using `gradlew :concept:tim-from-marketing-2:test`. Can you see why these test may fail?

josealonso commented:
> > The value of concepts.uuid is a zero-length string.
> > I thought it had to match the exercises.uuid, but doing that produced an error, so I left it empty.
> 
> They are two different UUIDs. You can generate one using `configlet uuid`.
> 
> > The tests are failing to build even locally, it might be a configuration issue.
> 
> I'm a little confused as to which solution to look at as there is one in `src/reference/java` and I can also see an implementation in `src/main/java`. I assume the one in `src/main/java` is just a work in progress (ultimately, we want to end up with the working one in `src/referece/java`). Could you let me know how you are currently testing?
Oh, sorry, I thought both files had the same content. I changed only the file name, as you suggested. I might have changed something else without realizing it. Let me take a look tonight (in Europe). 
Thanks for the hint, because I couldn't run the tests locally.

> Btw, to test against the reference solution, you can go into the `exercises` directory and run `./gradlew :concept:tim-from-marketing-2:test`.
Thanks, hopefully this command will work locally!

kahgoh commented on exercises/concept/tim-from-marketing-2/src/test/java/EmployeeDatabaseTest.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    @Test
+    @Tag("task:3")
```
There are only two tasks in the instructions, so the tags should be only `"task:1"` or `"task:2"`. They should line up with the corresponding task in the instructions.

kahgoh commented on concepts/optional-type/links.json:
```diff
@@ -0,0 +1,2 @@
+[]
```
Suggest a link to the Javadocs for Optional. See the links.json for the [Lists](https://github.com/exercism/java/blob/main/concepts/lists/links.json) concept for an example.

kahgoh commented:
No worries @josealonso. Thanks for trying and letting us know.
